llvm/lib/SYCLLowerIR/SYCLPropagateJointMatrixUsage.cpp

```diff
@@ -0,0 +1,237 @@
+    //   - sycl-joint-matrix-mad-type-A
+    //   - sycl-joint-matrix-mad-type-B
+    //   - sycl-joint-matrix-mad-type-C
+    //   - sycl-joint-matrix-mad-size-M
+    //   - sycl-joint-matrix-mad-size-K
+    //   - sycl-joint-matrix-mad-size-N
+    // NB! Values must be separated with a comma
+    Result +=
+        F->getFnAttribute("sycl-joint-matrix-mad-type-A").getValueAsString();
+    Result += ",";
+    Result +=
+        F->getFnAttribute("sycl-joint-matrix-mad-type-B").getValueAsString();
+    Result += ",";
+    Result +=
+        F->getFnAttribute("sycl-joint-matrix-mad-type-C").getValueAsString();
+    Result += ",";
+    Result +=
+        F->getFnAttribute("sycl-joint-matrix-mad-size-M").getValueAsString();
+    Result += ",";
+    Result +=
+        F->getFnAttribute("sycl-joint-matrix-mad-size-K").getValueAsString();
+    Result += ",";
+    Result +=
+        F->getFnAttribute("sycl-joint-matrix-mad-size-N").getValueAsString();
+    FunctionToJointMatrixMapValues[F].insert(Result);
+  }
+}
+
+/// Propagates joint matrix values from leaves up to the top of call graph.
+/// NB! Call graph corresponds to call graph of SYCL code which
+/// can't contain recursive calls. So there can't be loops in
+/// a call graph. But there can be path's intersections.
+void propagateJointMatrixValuesThroughCG(
+    Function *F, CallGraphTy &CG,
+    FunctionToJointMatrixValuesMapTy &FunctionToJointMatrixValues,
+    FunctionToJointMatrixValuesMapTy &FunctionToJointMatrixMadValues,
+    SmallPtrSet<const Function *, 16> &Visited) {
+  const auto It = CG.find(F);
+  if (It == CG.end())
+    return;
+
+  JointMatrixValuesSetTy LocalJointMatrixValues;
+  JointMatrixValuesSetTy LocalJointMatrixMadValues;
+  for (Function *Callee : It->second) {
+    if (Visited.insert(Callee).second)
+      propagateJointMatrixValuesThroughCG(
+          Callee, CG, FunctionToJointMatrixValues,
+          FunctionToJointMatrixMadValues, Visited);
+
+    const auto &CalleeJointMatrixValues = FunctionToJointMatrixValues[Callee];
+    LocalJointMatrixValues.insert(CalleeJointMatrixValues.begin(),
+                                  CalleeJointMatrixValues.end());
+    const auto &CalleeJointMatrixMadValues =
+        FunctionToJointMatrixMadValues[Callee];
+    LocalJointMatrixMadValues.insert(CalleeJointMatrixMadValues.begin(),
+                                     CalleeJointMatrixMadValues.end());
+  }
+  FunctionToJointMatrixValues[F].insert(LocalJointMatrixValues.begin(),
+                                        LocalJointMatrixValues.end());
+  FunctionToJointMatrixMadValues[F].insert(LocalJointMatrixMadValues.begin(),
+                                           LocalJointMatrixMadValues.end());
+}
+
+void setSyclJointMatrixMetadata(std::string MetadataName, Module *M,
+                                Function *F,
+                                FunctionToJointMatrixValuesMapTy ValuesMap) {
+  JointMatrixValuesSetTy Values = ValuesMap[F];
+  std::string StringValue;
```

Comment:
```suggestion
  SmallString<256> StringValue;
```

[Doxygen](https://llvm.org/doxygen/classllvm_1_1SmallString.html)
No need to re-allocate the whole string on each `+=`